bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java

```diff
@@ -661,4 +672,56 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
         }
     }
 
+    private ThingTypeRegistry thingTypeRegistry;
+    private ConfigDescriptionRegistry configDescriptionRegistry;
+
+    private void migrateThingType(Thing thing, ThingTypeUID thingTypeUID, Configuration configuration) {
+        ThingUID thingUID = thing.getUID();
+        ThingType thingType = thingTypeRegistry.getThingType(thingTypeUID);
+
+        // Remove the ThingHandler
+        final ThingHandlerFactory oldThingHandlerFactory = findThingHandlerFactory(thing.getThingTypeUID());
+        if (oldThingHandlerFactory != null) {
+            unregisterHandler(thing, oldThingHandlerFactory);
+        }
+
+        // Remove the links
+        thingLinkManager.thingRemoved(thing);
```

Comment:
since you are doing a thingUpdated() call in the end, you should not need to manually deal with the links here.